internal/pkg/debug/node/node.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package node contains utility functions for working with SSA nodes.
+package node
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/tools/go/ssa"
+)
+
+// CanonicalName produces a canonical string representation for an SSA node.
+func CanonicalName(n ssa.Node) string {
+	var target, op string
+	value, isValue := n.(ssa.Value)
+	instr, isInstr := n.(ssa.Instruction)
+	if isInstr {
+		op = instr.String()
+	}
+	if isValue {
+		if isInstr {
+			target = fmt.Sprintf("%s = ", value.Name())
+		} else {
+			target = value.Name()
+		}
+	}
+	return target + op
```

Comment:
A comment or example might be useful here.  I don't see why we don't need a separator between `target` and `op` for nodes which are both `Value` and `Instruction`, but I don't see anything tightly concatenated in the produced output.

Reply:
I think the concatenation works fine because:
* If a node is just an instr, target will be `""` so `target + op == op`.
* If a node is just a value, op will be `""` so `target + op == value.Name()`.
* If a node is both an instr and a value, target will have a space at the end, yielding `value.Name()<SPACE>instr.String()`.

Reply:
Now that I'm re-reading the code I agree that it could be clearer. I rewrote it without nesting and with explicit returns for each case, which I think makes it easier to follow. I added a default case for `ssa.Members` too.

WDYT?

Reply:
> If a node is both an instr and a value, target will have a space at the end [...]
That was what wasn't self-apparent to me.  That as a comment on the mixed case would be great.

Reply:
> I rewrote it without nesting and with explicit returns for each case, which I think makes it easier to follow. I added a default case for `ssa.Members` too.

Do you mean as a
```golang
switch {
 case isValue && isInstr:
 case isValue:
 case isInstr:
```

I was about to suggest that.

Reply:
Yep!